File: course_2/module_3/examples/red_black_tree.py
class Node:
    """Represents a node in the Red-Black Tree."""
    def __init__(self, key, color="red", parent=None, left=None, right=None):
        self.key = key  # The key value of the node.
        self.color = color  # The color of the node: "red" or "black".
        self.parent = parent  # Pointer to the parent node.
        self.left = left  # Pointer to the left child.
        self.right = right  # Pointer to the right child.


class RedBlackTree:
    """Implements a Red-Black Tree with standard operations."""
    def __init__(self):
        self.TNULL = Node(key=None, color="black")  # Sentinel node for leaves.
        self.root = self.TNULL  # Initialize the tree as empty.

    def search(self, key, node=None):
        """Searches for a key in the Red-Black Tree."""
        if node is None:  # If no starting node is specified, start at the root.
            node = self.root
        while node != self.TNULL and node.key != key:
            if key < node.key:  # Traverse to the left subtree.
                node = node.left
            else:  # Traverse to the right subtree.
                node = node.right
        return node  # Return the node if found, or TNULL if not found.

    def left_rotate(self, x):
        """Performs a left rotation to balance the tree."""
        y = x.right  # Set y as the right child of x.
        x.right = y.left  # Reassign the left subtree of y to the right of x.
        if y.left != self.TNULL:
            y.left.parent = x  # Update parent pointer for y's left subtree.
        y.parent = x.parent  # Update y's parent to x's parent.
        if x.parent is None:  # If x was the root, update the root.
            self.root = y
        elif x == x.parent.left:  # If x was a left child, update the parent's left pointer.
            x.parent.left = y
        else:  # Otherwise, update the parent's right pointer.
            x.parent.right = y
        y.left = x  # Make x the left child of y.
        x.parent = y  # Update x's parent pointer to y.

    def right_rotate(self, x):
        """Performs a right rotation to balance the tree."""
        y = x.left  # Set y as the left child of x.
        x.left = y.right  # Reassign the right subtree of y to the left of x.
        if y.right != self.TNULL:
            y.right.parent = x  # Update parent pointer for y's right subtree.
        y.parent = x.parent  # Update y's parent to x's parent.
        if x.parent is None:  # If x was the root, update the root.
            self.root = y
        elif x == x.parent.right:  # If x was a right child, update the parent's right pointer.
            x.parent.right = y
        else:  # Otherwise, update the parent's left pointer.
            x.parent.left = y
        y.right = x  # Make x the right child of y.
        x.parent = y  # Update x's parent pointer to y.

    def insert(self, key):
        """Inserts a new key into the Red-Black Tree."""
        new_node = Node(key, color="red", left=self.TNULL, right=self.TNULL)
        y = None  # Parent pointer.
        x = self.root  # Start at the root.

        while x != self.TNULL:  # Traverse the tree to find the insert location.
            y = x
            if new_node.key < x.key:
                x = x.left
            else:
                x = x.right

        new_node.parent = y  # Assign the parent.
        if y is None:  # If the tree was empty, make the new node the root.
            self.root = new_node
        elif new_node.key < y.key:  # Insert as the left child.
            y.left = new_node
        else:  # Insert as the right child.
            y.right = new_node

        if new_node.parent is None:  # If the new node is the root, color it black.
            new_node.color = "black"
            return

        if new_node.parent.parent is None:  # If the grandparent is None, nothing to fix.
            return

        self._fix_insert(new_node)  # Fix the tree to maintain Red-Black properties.

    def _fix_insert(self, node):
        """Fixes the tree after insertion to maintain Red-Black properties."""
        while node.parent and node.parent.color == "red":
            if node.parent == node.parent.parent.right:  # If parent is the right child.
                uncle = node.parent.parent.left  # Get the uncle node.
                if uncle.color == "red":  # Case 1: Uncle is red.
                    uncle.color = "black"
                    node.parent.color = "black"
                    node.parent.parent.color = "red"
                    node = node.parent.parent  # Move up the tree.
                else:
                    if node == node.parent.left:  # Case 2: Node is left child.
                        node = node.parent
                        self.right_rotate(node)
                    # Case 3: Node is right child.
                    node.parent.color = "black"
                    node.parent.parent.color = "red"
                    self.left_rotate(node.parent.parent)
            else:  # If parent is the left child.
                uncle = node.parent.parent.right  # Get the uncle node.
                if uncle.color == "red":  # Case 1: Uncle is red.
                    uncle.color = "black"
                    node.parent.color = "black"
                    node.parent.parent.color = "red"
                    node = node.parent.parent  # Move up the tree.
                else:
                    if node == node.parent.right:  # Case 2: Node is right child.
                        node = node.parent
                        self.left_rotate(node)
                    # Case 3: Node is left child.
                    node.parent.color = "black"
                    node.parent.parent.color = "red"
                    self.right_rotate(node.parent.parent)
            if node == self.root:  # If we reached the root, stop.
                break
        self.root.color = "black"  # Ensure the root is always black.

    def inorder_traversal(self, node=None, result=None):
        """Returns the in-order traversal of the tree."""
        if node is None:
            node = self.root  # Start at the root if no node is specified.
        if result is None:
            result = []  # Initialize the result list.
        if node != self.TNULL:  # Traverse the tree recursively.
            self.inorder_traversal(node.left, result)
            result.append(node.key)  # Append the current key.
            self.inorder_traversal(node.right, result)
        return result  # Return the sorted list of keys.


# Example usage
if __name__ == "__main__":
    rbt = RedBlackTree()

    # Insert elements.
    for key in [20, 15, 25, 10, 5, 30]:
        rbt.insert(key)

    # In-order traversal.
    print("In-order Traversal:", rbt.inorder_traversal())

    # Search for a key.
    search_key = 15
    found_node = rbt.search(search_key)
    if found_node != rbt.TNULL:
        print(f"Key {search_key} found with color {found_node.color}.")
    else:
        print(f"Key {search_key} not found.")

File: course_2/module_3/examples/red_black_tree.go
package main

import "fmt"

// Color constants for the Red-Black Tree nodes.
const (
	Red   = true
	Black = false
)

// Node represents a node in the Red-Black Tree.
type Node struct {
	Key    int   // The key of the node.
	Color  bool  // The color of the node: true for red, false for black.
	Parent *Node // Pointer to the parent node.
	Left   *Node // Pointer to the left child.
	Right  *Node // Pointer to the right child.
}

// RedBlackTree represents the Red-Black Tree.
type RedBlackTree struct {
	Root  *Node // Root node of the tree.
	TNULL *Node // Sentinel node (black).
}

// NewRedBlackTree initializes an empty Red-Black Tree.
func NewRedBlackTree() *RedBlackTree {
	tNull := &Node{Color: Black} // Initialize sentinel node.
	return &RedBlackTree{
		Root:  tNull,
		TNULL: tNull,
	}
}

// Search searches for a key in the Red-Black Tree.
func (rbt *RedBlackTree) Search(key int) *Node {
	current := rbt.Root
	for current != rbt.TNULL && key != current.Key {
		if key < current.Key {
			current = current.Left
		} else {
			current = current.Right
		}
	}
	return current
}

// LeftRotate performs a left rotation around the given node.
func (rbt *RedBlackTree) LeftRotate(x *Node) {
	y := x.Right
	x.Right = y.Left
	if y.Left != rbt.TNULL {
		y.Left.Parent = x
	}
	y.Parent = x.Parent
	if x.Parent == nil {
		rbt.Root = y
	} else if x == x.Parent.Left {
		x.Parent.Left = y
	} else {
		x.Parent.Right = y
	}
	y.Left = x
	x.Parent = y
}

// RightRotate performs a right rotation around the given node.
func (rbt *RedBlackTree) RightRotate(x *Node) {
	y := x.Left
	x.Left = y.Right
	if y.Right != rbt.TNULL {
		y.Right.Parent = x
	}
	y.Parent = x.Parent
	if x.Parent == nil {
		rbt.Root = y
	} else if x == x.Parent.Right {
		x.Parent.Right = y
	} else {
		x.Parent.Left = y
	}
	y.Right = x
	x.Parent = y
}

// Insert inserts a new key into the Red-Black Tree.
func (rbt *RedBlackTree) Insert(key int) {
	newNode := &Node{
		Key:    key,
		Color:  Red,
		Left:   rbt.TNULL,
		Right:  rbt.TNULL,
		Parent: nil,
	}

	y := (*Node)(nil)
	x := rbt.Root

	// Traverse the tree to find the correct position.
	for x != rbt.TNULL {
		y = x
		if newNode.Key < x.Key {
			x = x.Left
		} else {
			x = x.Right
		}
	}

	newNode.Parent = y
	if y == nil {
		rbt.Root = newNode
	} else if newNode.Key < y.Key {
		y.Left = newNode
	} else {
		y.Right = newNode
	}

	// If the new node is the root, color it black.
	if newNode.Parent == nil {
		newNode.Color = Black
		return
	}

	// If the grandparent is nil, no need to fix.
	if newNode.Parent.Parent == nil {
		return
	}

	rbt.fixInsert(newNode)
}

// fixInsert fixes the Red-Black Tree after an insertion.
func (rbt *RedBlackTree) fixInsert(node *Node) {
	for node.Parent != nil && node.Parent.Color == Red {
		if node.Parent == node.Parent.Parent.Right {
			uncle := node.Parent.Parent.Left
			if uncle != nil && uncle.Color == Red { // Case 1: Uncle is red.
				uncle.Color = Black
				node.Parent.Color = Black
				node.Parent.Parent.Color = Red
				node = node.Parent.Parent
			} else {
				if node == node.Parent.Left { // Case 2: Node is a left child.
					node = node.Parent
					rbt.RightRotate(node)
				}
				// Case 3: Node is a right child.
				node.Parent.Color = Black
				node.Parent.Parent.Color = Red
				rbt.LeftRotate(node.Parent.Parent)
			}
		} else {
			uncle := node.Parent.Parent.Right
			if uncle != nil && uncle.Color == Red { // Case 1: Uncle is red.
				uncle.Color = Black
				node.Parent.Color = Black
				node.Parent.Parent.Color = Red
				node = node.Parent.Parent
			} else {
				if node == node.Parent.Right { // Case 2: Node is a right child.
					node = node.Parent
					rbt.LeftRotate(node)
				}
				// Case 3: Node is a left child.
				node.Parent.Color = Black
				node.Parent.Parent.Color = Red
				rbt.RightRotate(node.Parent.Parent)
			}
		}
		if node == rbt.Root {
			break
		}
	}
	rbt.Root.Color = Black // Ensure the root is always black.
}

// GetInOrderTraversal returns the in-order traversal of the tree as a slice of keys.
func (rbt *RedBlackTree) GetInOrderTraversal() []int {
	var result []int
	var inorder func(node *Node)
	inorder = func(node *Node) {
		if node == rbt.TNULL {
			return
		}
		inorder(node.Left)
		result = append(result, node.Key)
		inorder(node.Right)
	}
	inorder(rbt.Root)
	return result
}

// Main function to demonstrate the Red-Black Tree.
func main() {
	rbt := NewRedBlackTree()

	// Insert elements.
	keys := []int{20, 15, 25, 10, 5, 30}
	for _, key := range keys {
		rbt.Insert(key)
	}

	// Print in-order traversal.
	fmt.Println("In-order Traversal:", rbt.GetInOrderTraversal())

	// Search for a key.
	searchKey := 15
	node := rbt.Search(searchKey)
	if node != rbt.TNULL {
		fmt.Printf("Key %d found with color %s.\n", searchKey, colorToString(node.Color))
	} else {
		fmt.Printf("Key %d not found.\n", searchKey)
	}
}

func colorToString(color bool) string {
	if color {
		return "Red"
	}
	return "Black"
}
